insights/client/crypto.py

```diff
@@ -0,0 +1,279 @@
+    """GPG command run in a temporary environment.
+
+    :param command: The command to be executed.
+    :type command: list[str]
+    :param key: Path to the GPG public key to check against.
+    :type key: str
+    :param _home: Path to the temporary GPG home directory.
+    :type _home: str | None
+    :param _raw_command: The last invoked command.
+    :type _raw_command: list[str] | None
+    """
+    TEMPORARY_GPG_HOME_PARENT_DIRECTORY = "/var/lib/insights/"
+
+    def __init__(self, command, key):
+        self.command = command
+        self.key = key
+        self._home = None
+        self._raw_command = None
+
+    def __str__(self):
+        return "<{cls} _home={home} _raw_command={raw}>".format(
+            cls=self.__class__.__name__,
+            home=self._home,
+            raw=self._raw_command
+        )
+
+    def _setup(self):
+        """Prepare GPG environment.
+
+        :returns: Result of the GPG key import.
+        :rtype: GPGCommandResult
+        """
+        # The /var/lib/insights/ directory is used instead of /tmp/ because
+        # GPG needs to have RW permissions in it, and existing SELinux rules only
+        # allow access here.
+        self._home = tempfile.mkdtemp(dir=type(self).TEMPORARY_GPG_HOME_PARENT_DIRECTORY)
+
+        logger.debug("setting up gpg temporary environment in '{home}'".format(home=self._home))
+        result = self._run(["--import", self.key])  # type: GPGCommandResult
+        if not result.ok:
+            logger.warning("failed to import key '{key}': {result}".format(
+                key=self.key, result=result
+            ))
+        return result
+
+    def _supports_cleanup_socket(self):
+        """Queries for the version of GPG binary.
+
+        :returns: `True` if the gnupg is known for supporting `--kill all`.
+        :rtype: bool
+        """
+        version_process = subprocess.Popen(
+            ["/usr/bin/gpg", "--version"],
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+            universal_newlines=True,
+            env={"GNUPGHOME": self._home},
+        )
+        stdout, stderr = version_process.communicate()
+        if version_process.returncode != 0:
+            stderr = "\n".join("stderr: {line}".format(line=line) for line in stderr.split("\n") if len(line))
+            logger.debug("could not query for gpg version:\n{err}".format(err=stderr))
+            return False
+
+        for line in stdout.split("\n"):
+            if line.startswith("gpg (GnuPG) "):
+                version = line.split(" ")[-1]  # type: str
+                break
+        else:
+            stdout = "\n".join("stdout: {line}".format(line=line) for line in stdout.split("\n") if len(line))
+            logger.debug("could not query for gpg version: output not recognized:\n{out}".format(out=stdout))
+            return False
+
+        try:
+            version_info = tuple(int(v) for v in version.split("."))
+        except Exception as exc:
```

Comment:
maybe one or more specific exceptions would be better than a broad `Exception` here

Reply:
There are multiple things that may go wrong there (`AttributeError`, `TypeError`, `ValueError`) depending on how malformed the line is. I think this is fine, we're logging the raw value when it fails.
I think it's fine.